### PR TITLE
Generation: fix handling of special tokens

### LINKED ISSUE #31251: Inconsistent behaviors of generate() between versions before and after 4.41.* on mistralai/Mistral-7B-Instruct-v0.2
### System Info

- `transformers` version: 4.41.1
- Platform: Linux-5.15.0-84-generic-x86_64-with-glibc2.31
- Python version: 3.9.19
- Huggingface_hub version: 0.23.1
- Safetensors version: 0.4.3
- Accelerate version: 0.30.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.3.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The following code has different behaviors before and after v4.41.* on mistralai/Mistral-7B-Instruct-v0.2, it may affect more models but I have only tested on this model.

```python
from importlib.metadata import version
model = AutoModelForCausalLM.from_pretrained('mistralai/Mistral-7B-Instruct-v0.2', device_map="auto")
tokenizer = AutoTokenizer.from_pretrained('mistralai/Mistral-7B-Instruct-v0.2')
messages = [{"role": "user", "content": "This is a test message"}]
messages = tokenizer.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
tokenizer.pad_token = tokenizer.eos_token
model_inputs = tokenizer(messages, return_tensors="pt", padding=True, add_special_tokens=False).to('cuda')

generation_config = GenerationConfig(do_sample=False, max_new_tokens=256, pad_token_id=tokenizer.eos_token_id)
generated_ids = model.generate(
    **model_inputs, generation_config=generation_config)[:, model_inputs["input_ids"].shape[1] :]

generated_texts = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
```

Furthermore, when I provide the generation configuration in the following way, the output is also not consistent in v4.41*:

```python
from transformers import GenerationConfig, AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained('mistralai/Mistral-7B-Instruct-v0.2', device_map="auto")
tokenizer = AutoTokenizer.from_pretrained('mistralai/Mistral-7B-Instruct-v0.2')
messages = [{"role": "user", "content": "This is a test message"}]
messages = tokenizer.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
tokenizer.pad_token = tokenizer.eos_token
model_inputs = tokenizer(messages, return_tensors="pt", padding=True, add_special_tokens=False).to('cuda')

# Differs here
generated_ids = model.generate(
    **model_inputs, max_new_tokens=256, do_sample=False, pad_token_id=tokenizer.eos_token_id)[:, model_inputs["input_ids"].shape[1] :]

generated_texts = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
```

### Expected behavior

In previous versions (e.g. 4.39.3), both ways of providing the configuration parameters will lead to the same generation:

> I see. If you have any specific question or request, feel free to ask and I'll do my best to help you out. Otherwise, this message will be considered as a test and no further action will be taken. Let me know if you need anything else.

But in v4.41.* ( e.g. 4.41.1), the first way will generate:

> I see. If you have any specific question or request, feel free to ask and I\'ll do my best to help you out. Otherwise, this message will be considered as a test and no further action will be taken. Let me know if you need anything else.\n\nHere are some examples of questions or requests I can help with:\n\n* "What\'s the weather like in New York City today?"\n* "Can you help me find a recipe for chocolate chip cookies?"\n* "How do I set up a new email account?"\n* "What\'s the best way to get from JFK airport to Times Square?"\n* "Can you recommend a good book to read?"\n\nLet me know if you have any other question or request, and I\'ll be happy to help. Otherwise, have a great day!\n\nIf you have any other question or request, feel free to ask and I\'ll do my best to help you out. Otherwise, this message will be considered as a test and no further action will be taken. Let me know if you need anything else.\n\nHere are some examples of questions or requests I can help with:\n\n* "What\'s the weather

while the second way will generate:

> I see. If you have any specific question or request, feel free to ask and I'll do my best to help you out. Otherwise, this message will be considered as a test and no further action will be taken. Let me know if you need anything else.

which is consistent with the older versions ( I suppose it's the expected behavior)

The behavior of `generate()` seems to be inconsistent both across versions and within 4.41.*. Or maybe I have misunderstood the documentation on how to provide the configuration parameters. Any help will be much appreciated!

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #31251. For backwards compatibility we have to check all special tokens firstly in kwargs, and then in self.config. Earlier we fixed that for `decoder_start_token_id`, this PR extends it for all special tokens.

Verified that it works with the provided script and the script from last fix on `decoder_start_token_id`

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
I had to fix the test by passing special token in config because previously all special tokens were updated inside generation strategy code blocks. I will see how to make it BC, not ready for merge yet

**Comment by zucchini-nlp** [comment]
I think it's better to just raise warning and ask users to provide attention masks, which I already commited. Another option would be to prepare special tokens twice, once in generate and the second time juts before pre-fill. 

I prefer the first options as it's less redundant code and more intuitive, lmk if we want full BC even with repeating code

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1497,6 +1487,15 @@ def _tensor_or_none(token, device=None):
             pad_token_id = eos_token_id[0]
             logger.warning(f"Setting `pad_token_id` to `eos_token_id`:{pad_token_id} for open-end generation.")
 
+        # we can't infer attn mask if pad token is set to be eos token in model's generation config
+        if eos_token_id is not None and torch.isin(elements=eos_token_id, test_elements=pad_token_id).any():
+            if kwargs_has_attention_mask is not None and not kwargs_has_attention_mask:
+                logger.warning(
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1493]: should we rather only warn once?

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1493]: right, will make once

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1463,25 +1446,32 @@ def _prepare_special_tokens(
...
-                generation_config.decoder_start_token_id, generation_config.bos_token_id
-            )
+        bos_token_id = _tensor_or_none(
+            generation_config.bos_token_id, self.generation_config.bos_token_id, device=device
+        )
+        eos_token_id = _tensor_or_none(
+            generation_config.eos_token_id, self.generation_config.eos_token_id, device=device
+        )
+        pad_token_id = _tensor_or_none(
+            generation_config.pad_token_id, self.generation_config.eos_token_id, device=device
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1466]: good catch!
